src/Build/Instance/ProjectItemInstance.cs

```diff
@@ -1386,9 +1386,10 @@ public void CopyMetadataTo(ITaskItem destinationItem, bool addOriginalItemSpec)
             /// <returns>The cloned metadata.</returns>
             public IDictionary CloneCustomMetadata()
             {
-                Dictionary<string, string> clonedMetadata = new Dictionary<string, string>(MSBuildNameIgnoreCaseComparer.Default);
+                var metadata = MetadataCollection;
```

Comment:
Are  you capturing the reference here to avoid data races between access to count and the enumerator?

Reply:
Calling MetadataCollection creates a new copy on each access, check the getter :)